Add OP-TEE support

lorc opened the pull request:
This PR adds OP-TEE support with virtualization mode to RPI5. It is quite convoluted process, because we need:

1. OP-TEE with RPI5 support 
2. Trusted Firmware-A with OP-TEE support
3. Some way to load OP-TEE binary into memory 
4. Integrate this in the Yocto build

 As it is impossible to force VPU loader to load two binaries, we glue TF-A and OP-TEE into one file named `armstub8-2712.bin`, then TF-A copies OP-TEE into designated memory region at the end of first GB.

lorc committed:
rwoptee: enable OP-TEE 4.2.0 build for raspberrypi5

Add OP-TEE 4.2.0 recipe appends tailored for "raspberrypi5"
machine. Patch with rpi5 port is being upstreamed right now and will
be removed when it will get into OP-TEE mainline.

As we are interested in Xen support, this recipe will build OP-TEE
with CFG_NS_VIRTUALIZATION enabled and support for two guests. Guest
count can be changed via local.conf by setting OPTEE_VIRT_GUEST_COUNT
variable.

Also, add dependency on meta-xt-security as we now using OP-TEE
recipes which are stored here.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc committed:
bsp: tf-a: add TF-A recipe with OP-TEE support

Recipe is based on similar recipe from meta-arm. We can't use meta-arm
directly because it relies on own set of toolchains and we don't want
to switch tool chain.

We are building latest (on time of writing) master branch for TF-A
becuse it got RPI5 support very recently and there is no official
releases that support RPI5.

This patch also enabled OP-TEE boot via TF-A. As a result of a build,
it will generate armstub8-2712.bin file, which is BL31+OP-TEE
combined. It then can be written to SD card to enable OP-TEE support
on the board.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc committed:
xen: enable OP-TEE mediator

We need OP-TEE support in our products, so enable OP-TEE mediator in
Xen hypervisor by setting CONFIG_TEE and CONFIG_OPTEE.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc committed:
image-minimal: add dependency on TF-A

Add dependency on TF-A, so it (together with OP-TEE OS) will be
deployed to SD card image. OP-TEE is dependency for TF-A, so we don't
need to add it manually.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

GrygiriiS commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-xen.dts:
```diff
@@ -29,6 +29,24 @@
 	};
+
+	/* Tell Xen about OP-TEE */
+	reserved-memory {
```
Hm.  Seems a bit entangled - may be we can re-use DT provided by FW? In boot script copy it where we need instead of loading from disk. (Yep static changes in dts will be lost)

oleksiimoisieiev commented on conf/layer.conf:
```diff
@@ -10,4 +10,4 @@ BBFILE_PATTERN_xt-rpi5 := "^${LAYERDIR}/"
 BBFILE_PRIORITY_xt-rpi5 = "10"
```
up streamed -> upstreamed

oleksiimoisieiev commented:
LGTM with a small comment. As for the changes to DT i think we can try to provide overlay which can be applied using config.txt commants. This will look neat. What do you think?
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

lorc commented:
> This will look neat. What do you think?

Great idea, I'll take a look.

lorc committed:
xen: dts: replace xen.dts with xen.dtso

Use device tree overlay instead of plain device tree. This has a
benefit that we will not override DTS passed from previous stage
bootloader.

Basically we are telling U-Boot to use device tree provided by
previous loader. U-Boot stores address for that DTB into ${fdt_addr}
variable. During boot process, we are loading xen-specific overlay
and apply it over the system DTB.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc committed:
xen: add TEE support for dom0less mode

With this patch it is possible to set TEE for a guest via "xen,tee"
property in the device tree.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc committed:
xen: dtso: enable OP-TEE for domd

This is needed mostly for testing purposes as right now we have only
one domain besides Zephyr itself.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>

lorc commented:
I added a patch that introduces DTB overlay: `xen: dts: replace xen.dts with xen.dtso`

GrygiriiS commented on recipes-kernel/linux/files/bcm2712-raspberrypi5-xen.dtso:
```diff
@@ -0,0 +1,63 @@
+	#address-cells = <0x2>;
+	#size-cells = <0x2>;
+	xen,xen-bootargs = "console=dtuart dtuart=/soc/serial@7d001000 dom0_mem=128M dom0_max_vcpus=1 xsm=flask flask=permissive";
+	xen,dom0-bootargs = "console=hvc0 earlycon=xen earlyprintk=xen clk_ignore_unused root=/dev/ram0";
```
could you drop bootargs here pls - they set in [xt-rpi-u-boot-scr.bb](https://github.com/xen-troops/meta-xt-rpi5/commit/bcfe5ce4107a178c6e435684f615fca071fb0df6#diff-bf65ca7b48fa2b6b3ea083f52969ca7e1ec6514ce23dd7c245ccc95774cec9cb)

lorc replied:
Yep, missed that. Should be fixed now

GrygiriiS commented:
Thank you `Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>`
